cachi2/interface/cli.py

```diff
@@ -405,6 +405,94 @@ def all_files_are_jsons(sbom_files_to_merge: Paths) -> Paths:
     return all_files_are_jsons(enough_files_for_merge(list(set(sbom_files_to_merge))))
 
 
+def merge_relationships(
+    relationships_list: List[List[SPDXRelation]], doc_ids: List[str], packages: List[SPDXPackage]
+) -> List[SPDXRelation]:
+    """Merge SPDX relationships.
+
+    Function takes two relationsships lists and unified list of packages.
+    For relationhips lists, map and inverse map of relations is created. SPDX document usually
+    contains virtual package which serves as "envelope" for all real packages. These virtual
+    packages are search in the relationsships and thier ID is stored as middle element.
+    """
+
+    def map_relationships(
+        relationships: List[SPDXRelation],
+    ) -> Tuple[Optional[str], Dict[str, List[str]], Dict[str, str]]:
+        relations_map: Dict[str, List[str]] = {}
+        inverse_map: Dict[str, str] = {}
+
+        for rel in relationships:
+            spdx_id, related_spdx = rel.spdxElementId, rel.relatedSpdxElement
+            relations_map.setdefault(spdx_id, []).append(related_spdx)
+            inverse_map[related_spdx] = spdx_id
+
+        root_element = next((k for k in relations_map if k not in inverse_map), None)
+        return root_element, relations_map, inverse_map
+
+    package_ids = {pkg.SPDXID for pkg in packages}
+    root_ids = []
+    maps = []
+    inv_maps = []
+    middles = []
+    for relationships, doc_id in zip(relationships_list, doc_ids):
+        root, _map, inv_map = map_relationships(relationships)
+        maps.append(_map)
+        inv_maps.append(inv_map)
+        if not root:
+            root = doc_id
+        root_ids.append(root)
+
+    for _map, _inv_map, root_id in zip(maps, inv_maps, root_ids):
+        middle = next((r for r, c in _map.items() if _inv_map.get(r) == root_id), None)
+        middles.append(middle)
+
+    merged_relationships = []
+
+    def process_relation(
+        rel: SPDXRelation,
+        root_main: Optional[str],
+        root_other: Optional[str],
+        middle_main: str,
+        middle_other: Optional[str],
+    ) -> None:
+        new_rel = SPDXRelation(
+            spdxElementId=root_main if rel.spdxElementId == root_other else rel.spdxElementId,
+            relatedSpdxElement=(
+                root_main if rel.relatedSpdxElement == root_other else rel.relatedSpdxElement
+            ),
+            relationshipType=rel.relationshipType,
+        )
+        if new_rel.spdxElementId == middle_other:
+            new_rel.spdxElementId = middle_main
+        if new_rel.spdxElementId in package_ids or new_rel.relatedSpdxElement in package_ids:
+            merged_relationships.append(new_rel)
+
+    middle_main = middles[0]
+    if not middle_main:
+        packages.append(
```

Comment:
`packages` is a mutable argument to a function. Here you mutate it and effectively create a second (hidden) return value. It is not clear if this is intentional or not. If it is then please redesign the function to return a tuple instead of mutating external objects in place. This is not a class method, the fact that `packages` will be mutated is not obvious without carefully reading the entire function. If it not intentional then please consider copying the list first.

Reply:
fixed